### PR TITLE
Introduce new script to automate the build process

### PR DESCRIPTION
**Pull request by ckelleyRH**
A new script `update_version.sh` is introduced, to start to reduce some of the manual burden of building packages. Usage is as:

`./update_version.sh <major> <minor> <update> <phase>`

...where `phase` is optional.

In this first iteration, it does the following:

* Updates the spec version to the new version provided
* Commits that change
* Creates a tag based on the new version provided
* Creates a source tarball based on the new version provided

Further updates to add resilience/further functionality will follow.

### CODE REVIEW

**Comment by ckelleyRH** [comment]
@edewata - the major thing I'm not sure of is updating the `release_number` variable:
```
# Downstream release number:
# - development/stabilization (unsupported): 0.<n> where n >= 1
# - GA/update (supported): <n> where n >= 1
%global           release_number 0.1
```
Would you expect to change this variable through this script? I have never altered it myself, but it is clearly meant to be linked with the `phase` variable which I do update.

**Comment by edewata** [comment]
@ckelleyRH I think so. We don't want to release incomplete code in the middle of development process into any Fedora versions because it could cause urgent problems for many people, so the build needs to be clearly marked with `0.<n>` release number and `alpha` phase. When we reach feature complete, we could release it in Rawhide with `0.<n>` release number and `beta` phase so people can try it but not use it in production. Once we reach code complete we could release it in the latest Fedora with the normal `<n>` release number (without any phase).

**Comment by edewata** [comment]
I think the scrirpt generally looks good, but we probably should provide an instruction on how to undo the tag or revert/amend the commit in case we find a problem after running the script.

**Comment by ckelleyRH** [comment]
> @ckelleyRH I think so. We don't want to release incomplete code in the middle of development process into any Fedora versions because it could cause urgent problems for many people, so the build needs to be clearly marked with `0.<n>` release number and `alpha` phase. When we reach feature complete, we could release it in Rawhide with `0.<n>` release number and `beta` phase so people can try it but not use it in production. Once we reach code complete we could release it in the latest Fedora with the normal `<n>` release number (without any phase).

OK, I'll have a think and do another iteration.

**Comment by ckelleyRH** [comment]
> I think the scrirpt generally looks good, but we probably should provide an instruction on how to undo the tag or revert/amend the commit in case we find a problem after running the script.

I actually have a script to do this, I ran `autobuild.sh` so many times to test i got bored of typing `git reset; git tag -d`😅 

I will include a fixed up version of that, and a readme, in the next iteration.

**Comment by ckelleyRH** [comment]
@edewata I addressed the following:

- Changing between alpha/beta and GA changes the `release_number` field.
- Renamed the script to better reflect what it does
- Added a check that the phase is a legit string
- Added some doc into the script
- Added a script that reverts the script, in case of :fire: 
- Updated the tests to check that the tag is created, if you get that far you've done OK.

Putting up for review again @fmarco76 @ladycfu @jmagne

File: .github/workflows/update-version-test.yml
Comment (edewata) [inline .github/workflows/update-version-test.yml:24]: To be more realistic, this should be `12.1.0-beta1`. We create alpha/beta releases before the GA which always ends with `.0`. After GA, we only release updates like `.1`, `.2`, etc. without any phases.

File: .github/workflows/update-version-test.yml
Comment (edewata) [inline .github/workflows/update-version-test.yml:31]: This can be changed to `12 1 0` to reflect the GA release.

File: .github/workflows/update-version-test.yml
Comment (edewata) [inline .github/workflows/update-version-test.yml:38]: We can use alpha/beta again in the next minor version before GA, i.e. `12.2.0-alpha1`.

**Review by edewata — COMMENTED** [review]
Thanks for the update! I have some comments about the test. I'm still looking at the script.

**Comment by ckelleyRH** [comment]
> Thanks for the update! I have some comments about the test. I'm still looking at the script.

Updated in next patchset, thanks!

**Comment by ckelleyRH** [comment]
OK, latest revision incorporates @edewata's suggestions, fixes a typo with the phase in the version and fixes the script so it actually passes the test (yay tests!).

File: update_version.sh
Comment (edewata) [inline update_version.sh:76]: Ideally if we update the phase we should bump the release number too so that the whole NVR can be sorted properly, but this should work just fine with `alpha<N>` and `beta<N>` phases as long as `N` is a single digit, so I'll let you decide if you want to do that.

If you choose not to bump the release number, we might as well just use `0.` prefix instead of `0.1` if the phase exists.

File: update_version.sh
Comment (edewata) [inline update_version.sh:59]: The conditions are probably not necessary. We can always update the version numbers regardless of the values, but I'll let you decide.

**Review by edewata — APPROVED** [review]
Thanks for the update! I have some comments below, but I think it LGTM. Feel free to update/merge.

File: update_version.sh
Comment (ckelleyRH) [reply update_version.sh:76]: Hmm, good point - I'll take a look.

Comment (edewata) [reply update_version.sh:76]: Just FYI, ideally the phase should be part of the version number, and the release should be just `0.N`:
* https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_complex_versioning
* https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/#_prerelease_versions

No need to implement this now, I think it will require more testing, but maybe in the future we can properly follow this guideline.

File: update_version.sh
Comment (ckelleyRH) [reply update_version.sh]: I removed the conditions, thanks!

File: update_version.sh
Comment (ckelleyRH) [reply update_version.sh]: I updated the script so now `release number` is:

- No `phase` = 1 (first release)
- `phase` updated = 0.<$CURRENT_RELEASE_NUMBER++>
-  `phase` added = 0.1 (first pre-release)

We can make further modifications later if we decide to change the versioning.

Comment (ckelleyRH) [reply update_version.sh]: > I updated the script so now `release number` is:
> 
> * No `phase` = 1 (first release)
> * `phase` updated = 0.<$CURRENT_RELEASE_NUMBER++>
> * `phase` added = 0.1 (first pre-release)
> 
> We can make further modifications later if we decide to change the versioning.

Hi @edewata, I saw you were OK with this merging but before I do would like to check that this update to the `release_number` makes sense to you, if you would be so kind as to re-review :smile:. Then I will merge and backport etc. Thanks!

Comment (edewata) [reply update_version.sh]: Yup, LGTM.

**Review by edewata — APPROVED** [review]

**Pull request merged**
